Import test: improve docs, add more tests (#73600)
ansible/ansible: docs/docsite/rst/dev_guide/testing/sanity/import.rst

import
======

All Python imports in the directories specified that are not from a limited set of possibilities listed below
must be imported in a try/except ImportError block as follows:

.. code-block:: python

    # Instead of 'import another_library', do:

    try:
        import another_library

>         HAS_ANOTHER_LIBRARY = True

Reviewer: For readability, it's probably better to have this in the `else:` block.

Author: Fixed in 4f9e1277be264add28921b348cdaf820b81a2a9f.

    except ImportError:
        HAS_ANOTHER_LIBRARY = False

    # Later in module code:

        if not HAS_ANOTHER_LIBRARY:
            # Needs: from ansible.module_utils.basic import missing_required_lib
            module.fail_json(msg=missing_required_lib('another_library'))

    # Later in plugin code:

        if not HAS_ANOTHER_LIBRARY:

>             raise AnsibleError('another_library must be installed to use this plugin')

Reviewer: I want to add that if the code targets a new enough Python 3 version, the example could preserve more metadata about the exception causes:
```python
    try:
        import another_library
    except ImportError as imp_exc:
        another_library_import_exc = imp_exc
    else:
        another_library_import_exc = None

    # Later in module code:

        if another_library_import_exc is not None:
            # Needs: from ansible.module_utils.basic import missing_required_lib
            module.fail_json(msg=missing_required_lib('another_library'))

    # Later in plugin code:

        if another_library_import_exc is not None:
            raise AnsibleError('another_library must be installed to use this plugin') from another_library_import_exc
```

For older Pythons the example could use `six.raise_from()` shim.

Author: I split up the example into modules vs. plugins, and used `traceback` to return the import traceback in `module.fail_json()`, and `six.raise_from()` for plugins (e74bec06995093c876c6fd81bc5615951641cba8).

import
======

> Ansible allows unchecked imports of some libraries from specific directories, listed at the bottom of this section. This allows Ansible to give better error messages to the user, and is a requirement for other sanity tests (such as ``validate-modules``). Import all other Python libraries in a try/except ImportError block as follows:

Reviewer 2: ```suggestion
Ansible allows unchecked imports of some libraries from specific directories, listed at the bottom of this section. Import all other Python libraries in a try/except ImportError block to support sanity tests such as ``validate-modules`` and to allow Ansible to give better error messages to the user. To import a library in a try/except ImportError block:
```

import
======

> All Python imports in the directories specified that are not from a limited set of possibilities listed below
> must be imported in a try/except ImportError block as follows:

Reviewer 2: Without the specific directories, I find this a little confusing. It's also passive voice. What about this?
```suggestion
Ansible allows unchecked imports of some libraries from specific directories, listed at the bottom of this section. Import all other Python libraries in a try/except ImportError block as follows:
```
Adding a short sentence about why this is useful/important/required would help make the rule more memorable.

Author: @acozine I added an explanation as well: 11d2f5f2a5ecbd6e3f117c7b7580597bc3751123

1. In modules:

   .. code-block:: python

       # Instead of 'import another_library', do:

       import traceback

       try:
           import another_library
       except ImportError:
           HAS_ANOTHER_LIBRARY = False
           ANOTHER_LIBRARY_IMPORT_ERROR = traceback.format_exc()
       else:
           HAS_ANOTHER_LIBRARY = True

       # Later in module code:

       module = AnsibleModule(...)

       if not HAS_ANOTHER_LIBRARY:
           # Needs: from ansible.module_utils.basic import missing_required_lib
           module.fail_json(
               msg=missing_required_lib('another_library'),
               exception=ANOTHER_LIBRARY_IMPORT_ERROR)

2. In plugins:

   .. code-block:: python

       # Instead of 'import another_library', do:

       from ansible.module_utils.six import raise_from

       try:
           import another_library
       except ImportError as imp_exc:
           ANOTHER_LIBRARY_IMPORT_ERROR = imp_exc
       else:
           ANOTHER_LIBRARY_IMPORT_ERROR = None

>        # Later in plugin code:

Reviewer 3: ```suggestion
       # Later in plugin code:
```
let's specify a suitable place maybe, e.g. in `__init__`?:)

Author: @Andersson007 how about:
```suggestion
       # Later in plugin code, for example in __init__ of the plugin:
```

       if ANOTHER_LIBRARY_IMPORT_ERROR:
           raise_from(
               AnsibleError('another_library must be installed to use this plugin'),
               ANOTHER_LIBRARY_IMPORT_ERROR)
           # If you target only newer Python 3 versions, you can also use the
           # 'raise ... from ...' syntax.

> The following shows which unchecked imports are allowed from which directories:

Reviewer 2: ```suggestion
Ansible allows the following unchecked imports from these specific directories:
```
